src/main/webapp/table-manager.js

```diff
@@ -106,36 +114,93 @@ class TableManager {
     this.table.setView({'columns': viewableCols});
   }
 
-  hasNextPage() {
+  hasNextPageCached() {
     return (this.currPage + 1) * this.pageSize < this.baseDataTable.getNumberOfRows();
   }
+  hasNextPageRemote() {
+    return this.nextPageToken != BLANK_PAGE_TOKEN;
+  }
+
+  hasNextPage() {
+    return this.hasNextPageRemote() || this.hasNextPageCached();
+  }
+
 
-  onPageChange(properties) {
+  async onPageChange(properties) {
     const pageDelta = properties['pageDelta']; // 1 or -1
     const newPage = this.currPage + pageDelta;
 
-    if (newPage < 0) {
+    if (newPage < 0 || (pageDelta > 0 && !this.hasNextPage()) ) {
       return;
     }
 
+    // Check if we have the necessary data to change the page.  If we dont, get it.
+    if (pageDelta > 0 && !this.hasNextPageCached()) {
+      await new Loading(this.addDataToTable.bind(this), true).load();
+    }
     this.currPage = newPage;
-    // TODO: Eventually this should use a server-side pagination endpoint to request material
     this.setDataTable();
 
     this.draw();
   }
 
-  onPageSizeChange(newPageSize) {
+  async onPageSizeChange(newPageSize) {
     this.pageSize = newPageSize;
     this.table.setOption('pageSize', newPageSize);
-    this.currPage = 0;
 
-    // TODO: Eventually this should use a server-side pagination endpoint to request material
-    this.setDataTable();
+    await this.updateNormal();
 
     this.draw();
   }
 
+  async resetNormalData() {
+    this.currPage = 0;
+    this.nextPageToken = BLANK_PAGE_TOKEN;
+
+    this.baseDataTable = new google.visualization.DataTable();
+    this.baseDataTable.addColumn('string', 'Serial Number');
+    this.baseDataTable.addColumn('string', 'Status');
+    this.baseDataTable.addColumn('string', 'Asset ID');
+    this.baseDataTable.addColumn('string', 'User');
+    this.baseDataTable.addColumn('string', 'Location');
+
+    let loader = new Loading(this.addDataToTable.bind(this), true);
+    await loader.load();
+  }
+
+  async addDataToTable() {
+    if (this.aggregating) {
+      throw new Error('Method addDataToTable should not be called while table is in aggregation mode');
+    }
+
+    let url = this.buildRequestURL();
+    await fetch(url)
+        .then(response => response.json())
+        .then(json => {
+          this.nextPageToken = json.nextPageToken == null ? BLANK_PAGE_TOKEN : json.nextPageToken;
+          for (let device of json.chromeosdevices) {
+            this.baseDataTable.addRow([
+                device.serialNumber,
+                device.status,
+                device.annotatedAssetId,
+                device.annotatedUser,
+                device.annotatedLocation]);
+          }
+        });
+  }
+
+  buildRequestURL() {
+    let url = new URL(PAGINATION_ENDPOINT, window.location.href);
```

Comment:
we use "url" a lot -- maybe rename to be more descriptive

Reply:
I've renamed `buildRequestURL` to be more descriptive (`buildNextPageURL`) so people will know what the url is referring to.